### ISSUE TITLE
Option parsing from environment variables

### ISSUE THREAD

**Issue by KyleLeneau** [apple/swift-argument-parser#4]
It would be super nice/helpful to parse an options value from an Environment variable value as well taking what is passed in the command line call as the priority over the ENV variable. For CLI's that run in CI or automated environments this is super helpful as it would allow not having to print the value in the ENV variable and could keep it hidden. Not only that but it would be very convenient for repeated uses on commands!

```swift
import ArgumentParser

struct MyCommand: ParsableCommand {
    @Option(name: .shortAndLong, help: "The API Secret.", env: "SERVICE_API_SECRET")
    var secret: String
}
```

**Usage:**
`mycommand --secret QWERTY12345`
or
`mycommand` (where `SERVICE_API_SECRET='QWERTY12345'`

**Comment by danieleggert** [comment]
Yes. We have been working on this already, and we’ll open a PR within the next week or two. There are still a lot of open questions about how to do this right, and we’d love to get feedback from everyone.

**Comment by KyleLeneau** [comment]
Awesome, I am happy to provide some feedback on a PR!

**Comment by KyleLeneau** [comment]
@danieleggert Any update on the progress of this?

**Comment by danieleggert** [comment]
> @danieleggert Any update on the progress of this?

Please check out #109 — it’s almost done.

**Comment by griffin-stewie** [comment]
any updates?

**Comment by kojirou1994** [comment]
still waiting😭

**Comment by philprime** [comment]
I this still on the roadmap?

**Comment by fkorotkov** [comment]
@rauhul, do you have a preference on @KyleLeneau's proposal in this issue vs @danieleggert implementation in https://github.com/apple/swift-argument-parser/pull/109 vs something new like auto-magically populate values based on command names and argument names. For example, if we have a root command `tart` and sub-command `push` with a `--populate-cache` flag then set it automatically to a value of `TART_PUSH_POPULATE_CACHE` if defined?

**Comment by rauhul** [comment]
I was thinking about a different approach where the parsing pipeline is composed of ArgumentValueProviders, the command line would be one such provider. I haven't fleshed out the idea but the benefit is that users could add their own argument providers depending on their use cases.

e.g. we could provide support out of the box for json config files, environment, `.env` files, NSUserDefaults, etc...

**Comment by manicmachine** [comment]
Any word on this? #109 got closed during the Swift branch transition and doesn't look like a replacement was created.

**Comment by kirilltitov** [comment]
So. Any progress on this one?

**Comment by kirilltitov** [comment]
Hello?

**Comment by rgoldberg** [comment]
@kirilltitov I'll probably integrate [Swift Configuration](https://github.com/apple/swift-configuration) with SAP in my own project.

Maybe SAP or SC could include my solution as a SwiftPM trait.

SC requires Swift 6.2+ & macOS 15+, though (not sure about other platforms).
